Mark set_ as an inplace view op

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115769

Summary: To be used in https://github.com/pytorch/pytorch/pull/113873. Since set_ is effectively an inplace view op, we'll need to skip caching them.

Test Plan: Built pytorch; specifically this step: `/home/slarsen/local/miniconda3/envs/pytorch-3.10/bin/python -m torchgen.gen --source-path /home/slarsen/local/pytorch/cmake/../aten/src/ATen --install_dir /home/slarsen/local/pytorch/build/aten/src/ATen --per-operator-headers --generate sources --output-dependencies /home/slarsen/local/pytorch/build/aten/src/ATen/generated_sources.cmake`

Differential Revision: [D52814561](https://our.internmc.facebook.com/intern/diff/D52814561)

masnesral committed:
Mark set_ as an inplace view op

Summary: To be used in https://github.com/pytorch/pytorch/pull/113873. Since set_ is effectively an inplace view op, we'll need to skip caching them.

Test Plan: Built pytorch; specifically this step: `/home/slarsen/local/miniconda3/envs/pytorch-3.10/bin/python -m torchgen.gen --source-path /home/slarsen/local/pytorch/cmake/../aten/src/ATen --install_dir /home/slarsen/local/pytorch/build/aten/src/ATen --per-operator-headers --generate sources --output-dependencies /home/slarsen/local/pytorch/build/aten/src/ATen/generated_sources.cmake`

[ghstack-poisoned]

masnesral commented on torchgen/gen_functionalization_type.py:
```diff
@@ -59,6 +59,8 @@
         # See Note [resize_ in Functionalization]
         "resize_",
         "resize_as_",
+        # set_ is a crazy op that swaps out storage
```
Probably there's a better comment. I'm just plagiarizing from messenger convo ;)

masnesral commented on torchgen/gen_functionalization_type.py:
```diff
@@ -776,7 +781,10 @@ def gen_functionalization_definition(
-        if str(g.func.name) not in MUTABLE_OPS_NOT_USING_FUNCTIONALIZATION:
+        if (
+            str(g.func.name) not in MUTABLE_OPS_NOT_USING_FUNCTIONALIZATION and
+            str(g.func.name.name) not in MUTABLE_OPS_NOT_USING_FUNCTIONALIZATION
```
Maybe better to just special case set_ here directly rather than adding it to `MUTABLE_OPS_NOT_USING_FUNCTIONALIZATION`?

bdhirsh replied:
tbh I think this is fine - we do the exact same thing with `resize_()` (where instead of having it go through all of this codegen, we just give it a dedicated, hand-written functionalization kernel here: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/FunctionalizeFallbackKernel.cpp#L329)

bdhirsh reviewed (approved):
looks great :) (if CI fails feel free to bug me and we can look around more)

masnesral committed:
lint on "Mark set_ as an inplace view op"


Summary: To be used in https://github.com/pytorch/pytorch/pull/113873. Since set_ is effectively an inplace view op, we'll need to skip caching them.

Test Plan: Built pytorch; specifically this step: `/home/slarsen/local/miniconda3/envs/pytorch-3.10/bin/python -m torchgen.gen --source-path /home/slarsen/local/pytorch/cmake/../aten/src/ATen --install_dir /home/slarsen/local/pytorch/build/aten/src/ATen --per-operator-headers --generate sources --output-dependencies /home/slarsen/local/pytorch/build/aten/src/ATen/generated_sources.cmake`

[ghstack-poisoned]

masnesral committed:
Update on "Mark set_ as an inplace view op"


Summary: To be used in https://github.com/pytorch/pytorch/pull/113873. Since set_ is effectively an inplace view op, we'll need to skip caching them.

Test Plan: Built pytorch; specifically this step: `/home/slarsen/local/miniconda3/envs/pytorch-3.10/bin/python -m torchgen.gen --source-path /home/slarsen/local/pytorch/cmake/../aten/src/ATen --install_dir /home/slarsen/local/pytorch/build/aten/src/ATen --per-operator-headers --generate sources --output-dependencies /home/slarsen/local/pytorch/build/aten/src/ATen/generated_sources.cmake`

[ghstack-poisoned]

bdhirsh commented on torchgen/gen_functionalization_type.py:
```diff
@@ -736,6 +738,9 @@ def emit_registration_helper(f: NativeFunction) -> str:
             return []
+        if str(f.func.name.name) == "set_":
+            # set_ is a crazy op that swaps out storage
+            return []
```
What happens if you remove this?

I think the problem is that we are no longer codegen'ing functionalziation kernels for any of the set_() ops, which we still want to do here

masnesral replied:
@bdhirsh I added this in an attempt to get around this build failure:

```
cd /home/slarsen/local/pytorch && /home/slarsen/local/miniconda3/envs/pytorch-3.10/bin/python -m torchgen.gen --source-path /home/slarsen/local/pytorch/cmake/../aten/src/ATen --install_dir /home/slarsen/local/pytorch/build/aten/src/ATen --per-operator-headers --generate sources --output-dependencies /home/slarsen/local/pytorch/build/aten/src/ATen/generated_sources.cmake
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/slarsen/local/pytorch/torchgen/gen.py", line 2880, in <module>
    main()
  File "/home/slarsen/local/pytorch/torchgen/gen.py", line 2820, in main
    gen_source_files(
  File "/home/slarsen/local/pytorch/torchgen/gen.py", line 2518, in gen_source_files
    cpu_fm.write_sharded(
  File "/home/slarsen/local/pytorch/torchgen/utils.py", line 237, in write_sharded
    env = env_callable(item)
          ^^^^^^^^^^^^^^^^^^
  File "/home/slarsen/local/pytorch/torchgen/gen.py", line 2490, in functionalization_env_callable
    "func_registrations": gen_functionalization_registration(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/slarsen/local/pytorch/torchgen/gen_functionalization_type.py", line 744, in gen_functionalization_registration
    assert not f.is_view_op
AssertionError
```

bdhirsh replied:
(hopefully) fixed offline

masnesral committed:
Update on "Mark set_ as an inplace view op"


Summary: To be used in https://github.com/pytorch/pytorch/pull/113873. Since set_ is effectively an inplace view op, we'll need to skip caching them.

Test Plan: Built pytorch; specifically this step: `/home/slarsen/local/miniconda3/envs/pytorch-3.10/bin/python -m torchgen.gen --source-path /home/slarsen/local/pytorch/cmake/../aten/src/ATen --install_dir /home/slarsen/local/pytorch/build/aten/src/ATen --per-operator-headers --generate sources --output-dependencies /home/slarsen/local/pytorch/build/aten/src/ATen/generated_sources.cmake`

[ghstack-poisoned]

masnesral committed:
Update on "Mark set_ as an inplace view op"


Summary: To be used in https://github.com/pytorch/pytorch/pull/113873. Since set_ is effectively an inplace view op, we'll need to skip caching them.

Test Plan: Built pytorch; specifically this step: `/home/slarsen/local/miniconda3/envs/pytorch-3.10/bin/python -m torchgen.gen --source-path /home/slarsen/local/pytorch/cmake/../aten/src/ATen --install_dir /home/slarsen/local/pytorch/build/aten/src/ATen --per-operator-headers --generate sources --output-dependencies /home/slarsen/local/pytorch/build/aten/src/ATen/generated_sources.cmake`

[ghstack-poisoned]

masnesral committed:
Update on "Mark set_ as an inplace view op"


Summary: To be used in https://github.com/pytorch/pytorch/pull/113873. Since set_ is effectively an inplace view op, we'll need to skip caching them.

Test Plan: Built pytorch; specifically this step: `/home/slarsen/local/miniconda3/envs/pytorch-3.10/bin/python -m torchgen.gen --source-path /home/slarsen/local/pytorch/cmake/../aten/src/ATen --install_dir /home/slarsen/local/pytorch/build/aten/src/ATen --per-operator-headers --generate sources --output-dependencies /home/slarsen/local/pytorch/build/aten/src/ATen/generated_sources.cmake`

[ghstack-poisoned]

masnesral committed:
Update on "Mark set_ as an inplace view op"


Summary: To be used in https://github.com/pytorch/pytorch/pull/113873. Since set_ is effectively an inplace view op, we'll need to skip caching them.

Test Plan: Built pytorch; specifically this step: `/home/slarsen/local/miniconda3/envs/pytorch-3.10/bin/python -m torchgen.gen --source-path /home/slarsen/local/pytorch/cmake/../aten/src/ATen --install_dir /home/slarsen/local/pytorch/build/aten/src/ATen --per-operator-headers --generate sources --output-dependencies /home/slarsen/local/pytorch/build/aten/src/ATen/generated_sources.cmake`

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge

jeanschmidt commented:
@pytorchbot revert -m "breaking landing signals internally, more details on the diff, author is tagged" -c ghfirst

masnesral committed:
Update on "Mark set_ as an inplace view op"


Summary: To be used in https://github.com/pytorch/pytorch/pull/113873. Since set_ is effectively an inplace view op, we'll need to skip caching them.

Test Plan: Built pytorch; specifically this step: `/home/slarsen/local/miniconda3/envs/pytorch-3.10/bin/python -m torchgen.gen --source-path /home/slarsen/local/pytorch/cmake/../aten/src/ATen --install_dir /home/slarsen/local/pytorch/build/aten/src/ATen --per-operator-headers --generate sources --output-dependencies /home/slarsen/local/pytorch/build/aten/src/ATen/generated_sources.cmake`

[ghstack-poisoned]

masnesral commented:
@masnesral has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52814561).

masnesral commented:
@pytorchmergebot

masnesral commented:
@pytorchbot merge
